File: Stack/225.Implement-Stack-using-Queues/225.Implement-Stack-using-Queues_v1.ts
class MyStack {
  private queue1: number[];
  private queue2: number[];
  private isInsertQ1: boolean;
  constructor() {
    this.queue1 = [];
    this.queue2 = [];
    this.isInsertQ1 = true;
  }

  push(x: number): void {
    if (this.isInsertQ1) {
      this.queue1.push(x);
    } else {
      this.queue2.push(x);
    }
  }

  pop(): number {
    this.move();
    if (this.isInsertQ1) {
      this.isInsertQ1 = false;
      return this.queue1.shift();
    } else {
      this.isInsertQ1 = true;
      return this.queue2.shift();
    }
  }

  top(): number {
    this.move();
    if (this.isInsertQ1) {
      return this.queue1[0];
    } else {
      return this.queue2[0];
    }
  }

  empty(): boolean {
    return this.queue1.length === 0 && this.queue2.length === 0;
  }

  move(): void {
    if (this.isInsertQ1) {
      while (this.queue1.length > 1) {
        this.queue2.push(this.queue1.shift());
      }
    } else {
      while (this.queue2.length > 1) {
        this.queue1.push(this.queue2.shift());
      }
    }
  }
}

/**
 * Your MyStack object will be instantiated and called as such:
 * var obj = new MyStack()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.empty()
 */

File: Stack/225.Implement-Stack-using-Queues/225.Implement-Stack-using-Queues_v1.py
class MyStack:

    def __init__(self):
        self.queue1 = []
        self.queue2 = []


    def push(self, x: int) -> None:
        if len(self.queue1) == 0:
            self.queue2.append(x)
        else:
            self.queue1.append(x)


    def pop(self) -> int:
        emptyQueue = None
        notEmptyQueue = None
        if len(self.queue1) == 0:
            emptyQueue = self.queue1
            notEmptyQueue = self.queue2
        else:
            emptyQueue = self.queue2
            notEmptyQueue = self.queue1
        
        while len(notEmptyQueue) > 1:
            emptyQueue.append(notEmptyQueue.pop(0))
        return notEmptyQueue.pop(0)

    def top(self) -> int:
        result = self.pop()
        self.push(result)
        return result


    def empty(self) -> bool:
        return len(self.queue1) == 0 and len(self.queue2) == 0



# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()
